# README.md

## Задание 1 - Анализ мест в коде с исключительными ситуациями

Можно вместо образца борда по ссылке в начале задания отлавливать исключения (errors) в функционале вашего калькулятора из предыдущих лабораторных работ, особенно в той, где надо было читать все необходимые и выраженния для вычислений из файла.

Исключительная ситуация может возникнуть на этапе работы с файлом (чтение, запись). Программа может не считать файл с настройками например, из-за ограниченного набора прав пользователя, запустившего данную программу или каких-либо настроек других ОС. Эти ситуации мы можем обработать с помощью блока (см. рабочий пример в стартовом борде):

```
try
    pass # какое-то выражение, возможно, поднимающее исключение  
except Exception:
   print('Исключение возникло') # обработка исключения 
else:
   # блок, выполняющийся, если исключения не было
finally:
   # блок, который выполняется в всегда независимо от того, было ли исключение или нет
```

**Комментарий:** 

**Результат программы:**

![code1]()

---

## Задание 2 - Модульное тестирование с unittest

Можно вместо образца борда по ссылке в начале задания покрыть тестами функционал вашего калькулятора из предыдущих лабораторных работ.

Шаблон для тестирования с помощью unittest может выглядеть так:

```
import unittest

class TestSomeFunc(unittest.TestCase): # создаем свой класс для тестов

     def firsttestcase(self): # внутри функции один или несколько тестовых
         self.assertEqual(2*2, 4) # случаев, которые проверяют какие-то 
         # близкие предположения
         

     # ...
     def secondtestcase(self): # вторая группа тестов
         pass


unittest.main(verbosity=1)     # запуск тестов
```

Пример тестирования двух функций convert_precision и two_sum, которую мы создавали ранее. Нюанс тестирования в repl.it и PyCharm. В repl.it  тесты запускаются вручную с помощью вкладки Shell (справа) (пример борда), в PyCharm требуется закомментировать запуск тестов с помощью: 

```
unittest.main(verbosity=1)
```


**Комментарий:** 

**Результат программы:**

![code2]()

---

## Задание 3 - Тестирование с помощью pytest

Можно вместо образца борда по ссылке в начале задания покрыть тестами функционал вашего калькулятора из предыдущих лабораторных работ.

По аналогии с предыдущим пунктом 2.

Перепешите те же самые тесты с помощью фреймворка pytest. Сравните его использование с фреймворком unittest.

Разберите работу с ним с помощью слендующих внешних источников:

    https://pytest-docs-ru.readthedocs.io/ru/latest/usage.html
    https://tproger.ru/articles/testiruem-na-python-unittest-i-pytest-instrukcija-dlja-nachinajushhih
    https://proglib.io/p/python-testing
    https://habr.com/ru/articles/448782/
    https://replit.com/talk/learn/Python-unit-tests-with-PyTest/40976
    https://docs.replit.com/teams-edu/unit-testing

**Комментарий:**

**Результат программы:**

![code3]()

---
